File: java/src/main/java/sk/epostak/sdk/resources/ConnectorWebhookResource.java
package sk.epostak.sdk.resources;

import sk.epostak.sdk.HttpClient;
import sk.epostak.sdk.models.ConnectorWebhookConfiguration;
import sk.epostak.sdk.models.ConnectorWebhookDeliveriesResponse;
import sk.epostak.sdk.models.ConnectorWebhookDeliveryDetail;
import sk.epostak.sdk.models.ConnectorWebhookDebuggerDeliveriesResponse;
import sk.epostak.sdk.models.ConnectorWebhookReplayResult;
import sk.epostak.sdk.models.ConnectorWebhookTestResponse;
import sk.epostak.sdk.models.ConnectorWebhookTestSuiteAccepted;
import sk.epostak.sdk.models.ConnectorWebhookTestSuiteStatus;
import sk.epostak.sdk.models.WebhookRotateSecretResponse;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/** One global Connector webhook per integrator. Every call omits X-Firm-Id. */
public final class ConnectorWebhookResource {
    private final HttpClient http;

    ConnectorWebhookResource(HttpClient http) {
        this.http = http;
    }

    public ConnectorWebhookConfiguration get() {
        return http.getNoFirm("/connector/webhook", ConnectorWebhookConfiguration.class);
    }

    public ConnectorWebhookConfiguration configure(String url, List<String> events) {
        if (url == null || url.isBlank()) {
            throw new IllegalArgumentException("Connector webhook URL is required");
        }
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("url", url.trim());
        if (events != null) body.put("events", List.copyOf(events));
        return http.putNoFirm("/connector/webhook", body, ConnectorWebhookConfiguration.class);
    }

    public ConnectorWebhookConfiguration configure(String url) {
        return configure(url, null);
    }

    public void delete() {
        http.deleteNoFirm("/connector/webhook", Void.class);
    }

    public WebhookRotateSecretResponse rotateSecret() {
        return http.postNoFirm(
                "/connector/webhook/rotate-secret",
                null,
                WebhookRotateSecretResponse.class
        );
    }

    public ConnectorWebhookTestResponse test(String customerRef) {
        return test(customerRef, null, null);
    }

    public ConnectorWebhookTestResponse test(String customerRef, String event, String scenario) {
        String normalizedCustomerRef = customerRef == null ? "" : ConnectorResource.trimString(customerRef);
        if (normalizedCustomerRef.isEmpty()) {
            throw new IllegalArgumentException("Connector customerRef is required");
        }
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("customerRef", normalizedCustomerRef);
        if (event != null) body.put("event", event);
        if (scenario != null) body.put("scenario", scenario);
        return http.postNoFirm(
                "/connector/webhook/test",
                body,
                ConnectorWebhookTestResponse.class
        );
    }

    public ConnectorWebhookDeliveriesResponse deliveries(String cursor, Integer limit, String status) {
        Map<String, Object> params = new LinkedHashMap<>();
        params.put("cursor", cursor);
        params.put("limit", limit);
        params.put("status", status == null ? null : status.toUpperCase(Locale.ROOT));
        return http.getNoFirm(
                "/connector/webhook/deliveries" + HttpClient.buildQuery(params),
                ConnectorWebhookDeliveriesResponse.class
        );
    }

    public ConnectorWebhookDeliveriesResponse deliveries() {
        return deliveries(null, null, null);
    }

    public ConnectorWebhookDebuggerDeliveriesResponse listDeliveries(Map<String, Object> filters) {
        Map<String, Object> params = new LinkedHashMap<>(filters == null ? Map.of() : filters);
        Object status = params.get("status");
        if (status != null) params.put("status", String.valueOf(status).toUpperCase(Locale.ROOT));
        return http.getNoFirm(
                "/connector/webhook/deliveries" + HttpClient.buildQuery(params),
                ConnectorWebhookDebuggerDeliveriesResponse.class
        );
    }

    public ConnectorWebhookDeliveryDetail getDelivery(String deliveryId) {
        return http.getNoFirm(
                "/connector/webhook/deliveries/" + HttpClient.encode(deliveryId),
                ConnectorWebhookDeliveryDetail.class
        );
    }

    public ConnectorWebhookReplayResult replayDelivery(
            String deliveryId,
            String idempotencyKey,
            boolean confirmSuccessfulReplay
    ) {
        if (idempotencyKey == null || idempotencyKey.isBlank()) {
            throw new IllegalArgumentException("Connector replay idempotencyKey is required");
        }
        return http.postIdempotentNoFirm(
                "/connector/webhook/deliveries/" + HttpClient.encode(deliveryId) + "/replay",
                Map.of("confirmSuccessfulReplay", confirmSuccessfulReplay),
                ConnectorWebhookReplayResult.class,
                idempotencyKey.trim()
        );
    }

    public ConnectorWebhookTestSuiteAccepted runTestSuite(
            String customerRef,
            String event,
            List<String> scenarios,
            String idempotencyKey
    ) {
        String normalizedCustomerRef = customerRef == null ? "" : ConnectorResource.trimString(customerRef);
        if (normalizedCustomerRef.isEmpty()) {
            throw new IllegalArgumentException("Connector customerRef is required");
        }
        if (idempotencyKey == null || idempotencyKey.isBlank()) {
            throw new IllegalArgumentException("Connector test-suite idempotencyKey is required");
        }
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("customerRef", normalizedCustomerRef);
        if (event != null) body.put("event", event);
        if (scenarios != null) body.put("scenarios", List.copyOf(scenarios));
        return http.postIdempotentNoFirm(
                "/connector/webhook/test-suite",
                body,
                ConnectorWebhookTestSuiteAccepted.class,
                idempotencyKey.trim()
        );
    }

    public ConnectorWebhookTestSuiteStatus getTestSuite(String testRunId) {
        return http.getNoFirm(
                "/connector/webhook/test-suite/" + HttpClient.encode(testRunId),
                ConnectorWebhookTestSuiteStatus.class
        );
    }
}

File: dotnet/src/EPostak/Resources/ConnectorWebhookResource.cs
using EPostak.Models;

namespace EPostak.Resources;

/// <summary>One global Connector webhook per integrator. Every call omits X-Firm-Id.</summary>
public sealed class ConnectorWebhookResource
{
    private readonly HttpRequestor _http;

    internal ConnectorWebhookResource(HttpRequestor http) => _http = http;

    public Task<ConnectorWebhookConfiguration> GetAsync(CancellationToken ct = default)
        => _http.RequestAsync<ConnectorWebhookConfiguration>(HttpMethod.Get, "/connector/webhook", ct, omitFirmId: true);

    public Task<ConnectorWebhookConfiguration> ConfigureAsync(
        string url,
        IReadOnlyList<string>? events = null,
        CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(url))
            throw new ArgumentException("Connector webhook URL is required.", nameof(url));
        return _http.RequestAsync<ConnectorWebhookConfiguration>(
            HttpMethod.Put,
            "/connector/webhook",
            new { url = url.Trim(), events },
            ct,
            omitFirmId: true);
    }

    public Task DeleteAsync(CancellationToken ct = default)
        => _http.RequestVoidAsync(HttpMethod.Delete, "/connector/webhook", ct, omitFirmId: true);

    public Task<WebhookRotateSecretResponse> RotateSecretAsync(CancellationToken ct = default)
        => _http.RequestAsync<WebhookRotateSecretResponse>(
            HttpMethod.Post,
            "/connector/webhook/rotate-secret",
            ct,
            omitFirmId: true);

    public Task<ConnectorWebhookTestResponse> TestAsync(
        string customerRef,
        CancellationToken ct = default)
        => TestScenarioAsync(customerRef, null, null, ct);

    public Task<ConnectorWebhookTestResponse> TestScenarioAsync(
        string customerRef,
        string? @event = null,
        string? scenario = null,
        CancellationToken ct = default)
    {
        var normalizedCustomerRef = ConnectorResource.TrimString(customerRef ?? "");
        if (normalizedCustomerRef.Length == 0)
            throw new ArgumentException("Connector customerRef is required.", nameof(customerRef));
        return _http.RequestAsync<ConnectorWebhookTestResponse>(
            HttpMethod.Post,
            "/connector/webhook/test",
            new Dictionary<string, object?>
            {
                ["customerRef"] = normalizedCustomerRef,
                ["event"] = @event,
                ["scenario"] = scenario,
            },
            ct,
            omitFirmId: true);
    }

    public Task<ConnectorWebhookDeliveriesResponse> DeliveriesAsync(
        string? cursor = null,
        int? limit = null,
        string? status = null,
        CancellationToken ct = default)
        => ListDeliveriesAsync(cursor, limit, status, ct: ct);

    public Task<ConnectorWebhookDeliveriesResponse> ListDeliveriesAsync(
        string? cursor = null,
        int? limit = null,
        string? status = null,
        string? customerRef = null,
        string? type = null,
        bool? test = null,
        string? from = null,
        string? to = null,
        CancellationToken ct = default)
    {
        var query = HttpRequestor.BuildQuery(
            ("cursor", cursor),
            ("limit", limit?.ToString()),
            ("status", status?.ToUpperInvariant()),
            ("customerRef", customerRef),
            ("type", type),
            ("test", test?.ToString().ToLowerInvariant()),
            ("from", from),
            ("to", to));
        return _http.RequestAsync<ConnectorWebhookDeliveriesResponse>(
            HttpMethod.Get,
            $"/connector/webhook/deliveries{query}",
            ct,
            omitFirmId: true);
    }

    public Task<ConnectorWebhookDeliveryDetail> GetDeliveryAsync(
        string deliveryId,
        CancellationToken ct = default)
        => _http.RequestAsync<ConnectorWebhookDeliveryDetail>(
            HttpMethod.Get,
            $"/connector/webhook/deliveries/{Uri.EscapeDataString(deliveryId)}",
            ct,
            omitFirmId: true);

    public Task<ConnectorWebhookReplayResult> ReplayDeliveryAsync(
        string deliveryId,
        string idempotencyKey,
        bool confirmSuccessfulReplay = false,
        CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(idempotencyKey))
            throw new ArgumentException("Connector replay idempotencyKey is required.", nameof(idempotencyKey));
        return _http.RequestIdempotentAsync<ConnectorWebhookReplayResult>(
            HttpMethod.Post,
            $"/connector/webhook/deliveries/{Uri.EscapeDataString(deliveryId)}/replay",
            new { confirmSuccessfulReplay },
            idempotencyKey.Trim(),
            ct,
            omitFirmId: true);
    }

    public Task<ConnectorWebhookTestSuiteAccepted> RunTestSuiteAsync(
        string customerRef,
        string idempotencyKey,
        string? @event = null,
        IReadOnlyList<string>? scenarios = null,
        CancellationToken ct = default)
    {
        var normalizedCustomerRef = ConnectorResource.TrimString(customerRef ?? "");
        if (normalizedCustomerRef.Length == 0)
            throw new ArgumentException("Connector customerRef is required.", nameof(customerRef));
        if (string.IsNullOrWhiteSpace(idempotencyKey))
            throw new ArgumentException("Connector test-suite idempotencyKey is required.", nameof(idempotencyKey));
        return _http.RequestIdempotentAsync<ConnectorWebhookTestSuiteAccepted>(
            HttpMethod.Post,
            "/connector/webhook/test-suite",
            new { customerRef = normalizedCustomerRef, @event, scenarios },
            idempotencyKey.Trim(),
            ct,
            omitFirmId: true);
    }

    public Task<ConnectorWebhookTestSuiteStatus> GetTestSuiteAsync(
        string testRunId,
        CancellationToken ct = default)
        => _http.RequestAsync<ConnectorWebhookTestSuiteStatus>(
            HttpMethod.Get,
            $"/connector/webhook/test-suite/{Uri.EscapeDataString(testRunId)}",
            ct,
            omitFirmId: true);
}
